blog/_posts/2019-09-05-put-down-particle-accelerator.md

```diff
@@ -0,0 +1,78 @@
+
+Build tools are particle accelerators. They're both large engineering projects that combine small pieces into progressively larger results. Compiling a C++ file with headers is like smashing protons and neutrons to create a Carbon atom. Linking many libraries to create a binary results in an Iron atom. Container images are even larger, like a Uranium atom. Want to combine files? Nothing's better.
+
+But what about when you your operation doesn't involve a file? If you need to run a server, what's the file that's output? How do you represent seeding a running database as a file operation? If your build failed because GitHub was down while it tried to download a dependency, how do you evict that entry from the cache? "Make everything as simple as possible, but no simpler." Files are too simple for microservice dev.
+
+## Development Today Needs New Tools
+For decades, the ways apps changed fit into make's model: projects got bigger, and thus build tools got faster. The move to microservices is different: it's a change in quality, not quantity.
+
+Microservices means your app isn't just one binary; your app (the experience your users care about) is the sum total of lots of individual binaries. Linking happens not at build-time but at run-time, over HTTP. Each individual build is smaller and faster. Starting the app is now more involved, as you have to connect all the components together in some way. You aren't delivering an atom; you're assembling a molecule.
+
+The new layer of working with molecules introduces new workflows. For example, `kubectl log` lets you get the independent stdout/stderr for each resource as it's running. Build tools don't automate this because it doesn't fit with their model, so it's up to the developer to make sure they're looking at the right log. It's easy to forget to check every nook and cranny, so developers end up missing an error and continue on a wild goose chase. Build tools are a hammer, and not everything's a nail.
+
+An example of a non-nail workflow: when I was at Google, I loved that our [monorepo](https://cacm.acm.org/magazines/2016/7/204032-why-google-stores-billions-of-lines-of-code-in-a-single-repository/fulltext) meant I could build and test any project with one command. It was cool that even though I never worked on GMail, I could refactor it easily. But if I wanted to actually change GMail, I'd have to go dig up a team-specific Wiki page or shell script that described how to start up the constituent pieces, spread across terminal tabs. Running your full app should be as easy as building a monolith.
```

Comment:
you've been leaning into the particle accelerator/chemistry set analogy (successfully, imo! that the sudden shift to hammer/nail is kind of jarring to me (esp the fact that it spans paragraphs so i need to keep it in my head instead of letting it be ephemeral -- it made sense in the prev paragraph, then i hit this one and had already forgotten it...)

Reply:
Thanks, removed the other analogy.